File: Sieve/sieve.c
/**
 * CIS-3190 A4 - Sieve of Erotasthenes in C
 * Author: Alex Lai
 * Student Number: 0920158
 * Email: alai02@uoguelph.ca
 * Date: Friday April 6 2018
**/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>

// Program to calcuate prime numbers until upper limit is reached
int main(void) {

    int maxPrime;

    printf("Sieve of Erotasthenes in C\n");

    // Ask for and accept user input
    printf("Enter an upper limit of primes: ");
    scanf("%d", &maxPrime);

    // Declare array of booleans with size entered from user
    bool isPrime[maxPrime + 1];
    memset(isPrime, true, sizeof(isPrime));
    
    // Call clock to get start time of algorithm 
    clock_t startTime = clock();
    
    // Start at 2 since 1 is not prime and loop until square root of upper limit is reached
    for (int i = 2; i * i <= maxPrime; i++) {
        // If array at i is prime, set all multiples to false
        if (isPrime[i]) {
            for (int j = i * 2; j <= maxPrime; j += i) {
                isPrime[j] = false;
            }
        }
    }
    
    // Call clock to get end time of the algorithm and calculate runtime
    clock_t endTime = clock();
    double execTime = (double)(endTime - startTime) / CLOCKS_PER_SEC;

    // Open file for writing and check for failure
    FILE* filePtr = fopen("c_sieve.txt", "w+");
    if (!filePtr) return EXIT_FAILURE;

    // Loop through array, writing prime valus to a file
    for (int i = 2; i <= maxPrime; i++) {
        if (isPrime[i]) {
            fprintf(filePtr, "%d\n", i);
        }
    }

    // Close the file and print out execution time to stdout
    fclose(filePtr);
    printf("Execution time: %lf seconds\n", execTime);

    return EXIT_SUCCESS;
}

File: Sieve/sieve.py

# CIS-3190 A4 - Sieve of Erotasthenes
# Author: Alex Lai
# Student Number: 0920158
# Email: alai02@uoguelph.ca
# Date: Friday April 6 2018

# Program to calcuate prime numbers until upper limit is reached
import math
import time

print 'Sieve of Erotasthenes in Python'

# Ask for and accept user input for upper limit
print 'Enter an upper limit of primes: '
maxPrime = raw_input()

# Set the input to a integer type
maxPrime = int(maxPrime)

# Initialzise array of boolean values with size of upper limit
isPrime = [False] * 2 + [True] * (maxPrime - 1) 

# Call time function for start time of the algorithm
startTime = time.time()

# Loop until square root of upper limit is reached
i = 2
while (i * i <= maxPrime):
    # Set multiples of prime values to false
    if isPrime[i]:
        for j in range(i * i, maxPrime + 1, i):
            isPrime[j] = False
    i = i + 1

# Call time function to get end time and compute execution time
endTime = time.time()
execTime = endTime - startTime

# Open file for output
file = open("py_sieve.txt","w+")

# Loop through array, writing prime values to file by converting them to strings
for i in xrange(maxPrime + 1):
    if isPrime[i]:
        file.write(str(i) + '\n')

# Close file and print out the execution time
file.close()
print 'Execution time: ', execTime, ' seconds'
